File: src/adapters/outbound/agents/investigation-agents.ts
import { Agent, type ModelSettings } from "@openai/agents";
import type {
	InvestigationContext,
	InvestigationRuntime,
} from "../../../ports/outbound/investigation-context";
import type { OpenAiSubAgentStreamCallbackFactory } from "./openai-subagent-stream-callback";
import { aggregateLogsByFacetTool } from "./tools/aggregate-datadog-logs-by-facet-tool";
import { getPullRequestDiffTool } from "./tools/get-pull-request-diff-tool";
import { getPullRequestTool } from "./tools/get-pull-request-tool";
import { getRepositoryContentTool } from "./tools/get-repository-content-tool";
import { listMetricsCatalogTool } from "./tools/list-datadog-metrics-catalog-tool";
import { queryMetricsTool } from "./tools/query-datadog-metrics-tool";
import { searchEventsTool } from "./tools/search-datadog-events-tool";
import { searchLogsTool } from "./tools/search-datadog-logs-tool";
import { createSearchGithubCodeTool } from "./tools/search-github-code-tool";
import { createSearchGithubIssuesAndPullRequestsTool } from "./tools/search-github-issues-and-pull-requests-tool";
import { createSearchGithubReposTool } from "./tools/search-github-repos-tool";

export type {
	InvestigationContext,
	InvestigationResources,
	InvestigationRuntime,
} from "../../../ports/outbound/investigation-context";

const INVESTIGATION_MODEL = "gpt-5.3-codex";

interface CoordinatorInstructionInput {
	datadogSite: string;
	githubScopeOrg: string;
	runtime: InvestigationRuntime;
	language: string;
}

interface AgentLanguageInput {
	language: string;
}

interface CreateCoordinatorAgentInput {
	logsAgent: Agent<InvestigationContext>;
	metricsAgent: Agent<InvestigationContext>;
	eventsAgent: Agent<InvestigationContext>;
	githubExplorerAgent: Agent<InvestigationContext>;
	onSubAgentStream?: OpenAiSubAgentStreamCallbackFactory;
	language: string;
}

export interface CreateSynthesizerAgentInput {
	language: string;
}

function buildCoordinatorInstructions(
	input: CoordinatorInstructionInput,
): string {
	return `
You are an SRE/Security/Platform engineer operating from Slack mentions. 

## Output Language
${input.language}

Language Policy:
- Use ${input.language} for all conversation and reasoning.
- Keep every response, plan, and analysis in ${input.language}.
- Do not switch languages unless explicitly requested by the user.

Current run context:
- Now: ${new Date().toISOString()}
- Slack Channel: ${input.runtime.channel}
- Slack Thread: ${input.runtime.threadTs}
- Retry Count: ${input.runtime.retryCount}
- GitHub Organization Scope: ${input.githubScopeOrg}

Your job is to orchestrate a thorough and fast investigation of requests that arrive via Slack mentions, such as:
- Another member’s investigation instruction (e.g., “please check this alert”, “triage this”, “estimate impact”, “find the likely cause”)
- Alert/notification investigation requests (full alert bodies, links, pasted logs, dashboards)
- Ad-hoc verification and diagnostic requests related to reliability, security, or platform operations

You must:
- Treat the Slack mention content as heterogeneous input (instructions, alert text, partial context, links, pasted logs, or free-form questions).
- Infer intent and required outcome from the received input, then decide what to do next (triage, scope, hypotheses, evidence collection, correlation, mitigation guidance, and what to report back).
- Take appropriate actions using the available tools, and delegate/coordinate investigation steps across specialized sub-agents when helpful, while remaining accountable for the overall outcome.
- Optimize for speed, correctness, and actionable results for the requesting member (not “incident response process” unless explicitly requested).

## Investigation workflow (MUST follow)
1. Classify the request before running tools:
  - Investigation mode: incident/alert triage, impact estimation, root-cause analysis, anomaly diagnosis, or reliability/security/platform verification.
  - Direct task mode: simple retrieval or lookup where full incident workflow is unnecessary (for example, listing GitHub repositories, fetching a specific PR/file, or answering a narrow factual question).
2A. If the request is Investigation mode and related to systems/services, first establish a system map from source-of-truth (GitHub) before querying observability:
  - Identify the relevant repository (from alert links, service name, owning team, or dependency hints). If ambiguous, pick the most likely and proceed; do not stall.
  - Read README and any architecture/runbook/oncall docs (e.g., /docs, /runbooks, /architecture, /ops).
  - Extract and summarize:
    - Service name(s), deploy artifacts, runtime (k8s/ecs/lambda), and environments (prod/stg/dev)
    - External dependencies (DB, queue, third-party APIs) and internal upstream/downstream services
    - Entry points (API routes, workers, cron), SLO/SLI, dashboards, and known failure modes
    - Ownership signals (CODEOWNERS, team labels, Slack channels, oncall rotation hints)
  - Search for recent changes likely related to the issue:
    - PRs merged, releases/tags, config changes (env vars, feature flags), infra changes (IaC) within the suspected time window
    - Prefer PRs touching critical paths (auth, routing, DB, cache, queue, rate limit, timeout, retry)
  - Record “system map” findings as hypotheses inputs (what changed, where to look in logs/metrics/events).
2B. Use aggregate_datadog_logs_by_facet with facet=service to get an overview of important service tags used for Datadog investigations.
  and then use list_datadog_metrics_catalog to see which metrics are active and infer what systems/services are currently running.
3. In Investigation mode, after building this context, take action for the specific input using tools.
4. In Direct task mode, skip unnecessary environment-wide investigation and execute the minimum tool calls required to answer quickly and accurately.

## Parallelism rules (MUST follow)
- Execute tool calls in parallel whenever possible; avoid sequential execution when requests are independent.
- In follow-up rounds, batch as many tool calls as possible in a single step. If two sources need deeper queries, issue both calls at the same time.
- Never wait for one tool result before starting another unrelated tool call.

## Follow-up strategy
After receiving the initial results, decide whether deeper investigation is needed:
- A log error mentions a specific service → re-query logs scoped to that service in parallel with any other needed queries.
- A metric spike aligns with a deployment timestamp → re-query events for that deploy in parallel with a narrower metric query.
- A source returns no data → try an alternative query for that source while continuing with others.
- If a tool returns {"ok":false,"kind":"client_error"}, treat it as a query/input issue, adjust the query from the message, and retry when useful.
Stop when you have sufficient evidence to understand the probable cause, or when additional queries return no new information.

## Final reporting requirements (MUST follow)
Choose the output style based on request mode:
- Investigation mode:
  - Produce a concise investigation report that captures multiple plausible findings, because investigations often yield several “this might be it” facts rather than a single definitive root cause.
  - Present the top suspected explanations as separate findings, each grounded in concrete evidence.
  - For each finding, include:
    - A short title (what it likely is)
    - Evidence summary (what you observed)
    - Why it matters / how it explains the request
    - Confidence (0 – 100 %)
    - Next action(s) to confirm or mitigate
  - Construct and include Datadog deep links (URLs) for every key Datadog evidence item so a human can click through:
    - Build links for logs queries (Log Explorer) using the exact query/time window used in investigation.
    - Build links for metrics graphs/dashboards (Metrics Explorer / relevant dashboard widgets) matching the time window and scope.
    - Build links for events/deploys/incidents (Event Stream / change events) matching the time window.
    - Ensure URLs include this Datadog site: ${input.datadogSite || "datadoghq.com"} and encode the query/time range so the view reproduces the evidence.
  - Include a short “What I checked” section describing tools and scopes used (services, env, time range), and a “What I did not find” section if relevant.
- Direct task mode:
  - Return a concise direct answer optimized for speed.
  - Include only the requested result and a brief execution summary (which tools/queries were used).
  - Do not force incident-style findings or Datadog deep links when Datadog investigation was not needed.
`;
}

function buildSynthesizerInstructions(input: AgentLanguageInput): string {
	return `You are an SRE/Security/Platform engineering expert. You receive structured investigation reports collected from multiple sources (logs, metrics, alerts, events, dashboards, etc.) and your job is to synthesize the findings into a clear, actionable Slack post for the team.

Your output is a Slack message — not an internal document. It should be easy to scan quickly, written in a tone that informs and prompts action without causing unnecessary panic.

## Output Language
${input.language}

## Language Policy
- Use ${input.language} for all conversation and reasoning.
- Keep every response and analysis in ${input.language}.

## Guidelines

- Write for a Slack audience: scannable, concise, no walls of text.
- Use Slack markdown (*bold*, bullet points). Do not use headers with \`##\` or \`**\` markdown — use \`*text*\` for bold instead.
- Only state what is supported by the provided report.
- Do not omit evidence links or traceability steps. For each important claim, include the supporting URL and a brief note on how to follow/verify it.
`;
}

function buildLogsInstructions(input: AgentLanguageInput): string {
	return `You are a log analysis specialist. Use the search_datadog_logs tool to investigate the given query.
Analyze the log results and provide a concise summary of what you found, including any errors, anomalies, or patterns.
Focus on timestamps, error messages, service names, and status codes.
When multiple independent tool calls are needed, run them in parallel whenever possible and only run sequentially when there is a strict dependency.
Always include a brief execution report that states which query you used for each tool call and what you did.
If a tool result contains {"ok":false,"kind":"client_error"}, treat it as a query/input issue, use the returned message to adjust the query, and retry when useful.

## Language Policy
Use ${input.language} for all conversation and reasoning.
Keep every response, plan, and analysis in ${input.language}.`;
}

function buildMetricsInstructions(input: AgentLanguageInput): string {
	return `You are a metrics analysis specialist. Use the query_datadog_metrics tool to investigate the given metric query.
Analyze the metric data points and provide a concise summary of trends, spikes, or anomalies you observe.
Focus on identifying unusual patterns that could indicate issues.
When multiple independent tool calls are needed, run them in parallel whenever possible and only run sequentially when there is a strict dependency.
Always include a brief execution report that states which query you used for each tool call and what you did.
If a tool result contains {"ok":false,"kind":"client_error"}, treat it as a query/input issue, use the returned message to adjust the query, and retry when useful.

## Language Policy
Use ${input.language} for all conversation and reasoning.
Keep every response, plan, and analysis in ${input.language}.`;
}

function buildEventsInstructions(input: AgentLanguageInput): string {
	return `You are an events analysis specialist. Use the search_datadog_events tool to investigate the given query.
Analyze the events and provide a concise summary focusing on deployments, configuration changes, GitHub activity,
and any other events that could be related to an incident.
When multiple independent tool calls are needed, run them in parallel whenever possible and only run sequentially when there is a strict dependency.
Always include a brief execution report that states which query you used for each tool call and what you did.
If a tool result contains {"ok":false,"kind":"client_error"}, treat it as a query/input issue, use the returned message to adjust the query, and retry when useful.

## Language Policy
Use ${input.language} for all conversation and reasoning.
Keep every response, plan, and analysis in ${input.language}.`;
}

const modelSettings: ModelSettings = {
	reasoning: {
		effort: "low",
		summary: "auto",
	},
	text: {
		verbosity: "low",
	},
	parallelToolCalls: true,
};

export function createLogsAgent(
	input: AgentLanguageInput,
): Agent<InvestigationContext> {
	return new Agent<InvestigationContext>({
		name: "LogsInvestigator",
		model: INVESTIGATION_MODEL,
		modelSettings,
		instructions: buildLogsInstructions({
			language: input.language,
		}),
		tools: [searchLogsTool],
	});
}

export function createMetricsAgent(
	input: AgentLanguageInput,
): Agent<InvestigationContext> {
	return new Agent<InvestigationContext>({
		name: "MetricsInvestigator",
		model: INVESTIGATION_MODEL,
		modelSettings,
		instructions: buildMetricsInstructions({
			language: input.language,
		}),
		tools: [queryMetricsTool],
	});
}

export function createEventsAgent(
	input: AgentLanguageInput,
): Agent<InvestigationContext> {
	return new Agent<InvestigationContext>({
		name: "EventsInvestigator",
		model: INVESTIGATION_MODEL,
		modelSettings,
		instructions: buildEventsInstructions({
			language: input.language,
		}),
		tools: [searchEventsTool],
	});
}

export function createGitHubExplorerAgent(
	input: AgentLanguageInput,
): Agent<InvestigationContext> {
	const searchGithubCodeTool = createSearchGithubCodeTool();
	const searchGithubReposTool = createSearchGithubReposTool();
	const searchGithubIssuesAndPullRequestsTool =
		createSearchGithubIssuesAndPullRequestsTool();

	return new Agent<InvestigationContext>({
		name: "GitHubExplorer",
		model: INVESTIGATION_MODEL,
		modelSettings,
		instructions: (
			context,
		) => `You are a GitHub code and repository analysis specialist.
Use the available tools to search code, repositories, issues/PRs, read files, and inspect pull requests inside the configured organization scope only.

GitHub org that must be specified in search queries: ${context.context.resources.githubScopeOrg}

Mandatory query rule:
- For every call to search_github_code, search_github_repos, and search_github_issues_and_pull_requests, explicitly include org:${context.context.resources.githubScopeOrg} in query.
- Never omit org qualifier in GitHub search queries.

Your investigation strategy:
- Use search_github_code to find specific files, configs, or code patterns across repositories.
- search_github_code has stricter rate limits than other GitHub searches. Prefer broader, high-recall keyword queries that can surface many relevant candidates per call, and avoid many narrow sequential code searches.
- Use search_github_repos to discover repositories by name, topic, or language.
- Use search_github_issues_and_pull_requests to find recent PRs, open issues, or merged changes related to a service or incident.
- Use get_repository_content to read specific files (configs, runbooks, source code) and check kind/truncated fields.
- Use get_pull_request to check metadata of a specific pull request.
- Use get_pull_request_diff sparingly and only after narrowing the scope. Handle truncated diff safely.

When multiple independent searches or reads are needed, run them in parallel.
Always include a brief execution report that states which queries/files you accessed and what you found.
If a tool result contains {"ok":false,"kind":"client_error"}, treat it as a query/input issue, use the returned message to adjust the query/target, and retry when useful.

## Language Policy
Use ${input.language} for all conversation and reasoning.
Keep every response, plan, and analysis in ${input.language}.`,
		tools: [
			searchGithubCodeTool,
			searchGithubReposTool,
			searchGithubIssuesAndPullRequestsTool,
			getRepositoryContentTool,
			getPullRequestTool,
			getPullRequestDiffTool,
		],
	});
}

export function createCoordinatorAgent(
	input: CreateCoordinatorAgentInput,
): Agent<InvestigationContext> {
	const subAgentsRunOptions = {
		maxTurns: 50,
	};
	return new Agent<InvestigationContext>({
		name: "Coordinator",
		model: INVESTIGATION_MODEL,
		instructions: (context) => {
			const investigationContext = context.context;
			return buildCoordinatorInstructions({
				datadogSite:
					investigationContext.resources.datadogSite || "datadoghq.com",
				githubScopeOrg: investigationContext.resources.githubScopeOrg,
				runtime: investigationContext.runtime,
				language: input.language,
			});
		},
		modelSettings: modelSettings,
		tools: [
			aggregateLogsByFacetTool,
			listMetricsCatalogTool,
			input.logsAgent.asTool({
				toolName: "investigate_logs",
				runOptions: { ...subAgentsRunOptions },
				toolDescription:
					"Delegate a log investigation to the Logs specialist. Describe your hypothesis, what you want to verify, and investigation scope (service/env/time window/symptoms). Ask the specialist to design and run the necessary Datadog log queries.",
				onStream: input.onSubAgentStream?.({
					parentToolCallId: "investigate_logs",
				}),
			}),
			input.metricsAgent.asTool({
				toolName: "investigate_metrics",
				runOptions: { ...subAgentsRunOptions },
				toolDescription:
					"Delegate a metrics investigation to the Metrics specialist. Describe your hypothesis, what you want to verify, and investigation scope (systems/signals/time window). Ask the specialist to choose and run the necessary Datadog metric queries.",
				onStream: input.onSubAgentStream?.({
					parentToolCallId: "investigate_metrics",
				}),
			}),
			input.eventsAgent.asTool({
				toolName: "investigate_events",
				runOptions: { ...subAgentsRunOptions },
				toolDescription:
					"Delegate an events investigation to the Events specialist. Describe your hypothesis, what you want to verify, and investigation scope (deployments/changes/repos/time window). Ask the specialist to select and run the necessary Datadog event searches.",
				onStream: input.onSubAgentStream?.({
					parentToolCallId: "investigate_events",
				}),
			}),
			input.githubExplorerAgent.asTool({
				toolName: "investigate_github",
				runOptions: { ...subAgentsRunOptions },
				toolDescription: `Delegate a GitHub investigation to the GitHub specialist. Describe what you want to find (service configs, runbooks, recent PRs, architecture docs, deployment manifests, etc.) and the relevant service/repo names if known. All GitHub search queries must explicitly include the configured org qualifier.`,
				onStream: input.onSubAgentStream?.({
					parentToolCallId: "investigate_github",
				}),
			}),
		],
		outputType: "text",
	});
}

export function createSynthesizerAgent(
	input: CreateSynthesizerAgentInput,
): Agent {
	return new Agent({
		name: "Synthesizer",
		model: INVESTIGATION_MODEL,
		instructions: () =>
			buildSynthesizerInstructions({
				language: input.language,
			}),
	});
}

File: reili/adapters/src/outbound/agents/investigation_agents.rs
use std::sync::Arc;

use chrono::Utc;
use reili_shared::ports::outbound::{
    COORDINATOR_PROGRESS_OWNER_ID, InvestigationProgressEventPort, InvestigationResources,
    InvestigationRuntime,
};
use reili_shared::types::AlertContext;
use rig::agent::Agent;
use rig::prelude::CompletionClient;
use rig::providers::openai;
use serde_json::json;

use super::tools::{
    AggregateDatadogLogsByFacetTool, GetPullRequestDiffTool, GetPullRequestTool,
    GetRepositoryContentTool, ListDatadogMetricsCatalogTool, QueryDatadogMetricsTool,
    ReportProgressTool, ReportProgressToolInput, SearchDatadogEventsTool, SearchDatadogLogsTool,
    SearchGithubCodeTool, SearchGithubIssuesAndPullRequestsTool, SearchGithubReposTool,
    SearchWebTool,
};
use super::{
    progress_event_hook::ProgressEventHook,
    progress_reporting_sub_agent_tool::ProgressReportingSubAgentTool,
};

const INVESTIGATION_MODEL: &str = "gpt-5.3-codex";
const COORDINATOR_MAX_TURNS: usize = 20;
const SUBAGENT_MAX_TURNS: usize = 50;
const LOGS_PROGRESS_OWNER_ID: &str = "investigate_logs";
const METRICS_PROGRESS_OWNER_ID: &str = "investigate_metrics";
const EVENTS_PROGRESS_OWNER_ID: &str = "investigate_events";
const GITHUB_PROGRESS_OWNER_ID: &str = "investigate_github";

type OpenAiCompletionModel = <openai::Client as CompletionClient>::CompletionModel;
type OpenAiAgent = Agent<OpenAiCompletionModel>;
type OpenAiSubAgent = Agent<OpenAiCompletionModel, ProgressEventHook>;

pub struct BuildCoordinatorAgentInput {
    pub client: openai::Client,
    pub resources: Arc<InvestigationResources>,
    pub datadog_site: String,
    pub github_scope_org: String,
    pub runtime: InvestigationRuntime,
    pub on_progress_event: Arc<dyn InvestigationProgressEventPort>,
    pub language: String,
}

#[must_use]
pub fn build_coordinator_agent(input: BuildCoordinatorAgentInput) -> OpenAiAgent {
    let logs_agent = build_logs_agent(BuildSpecialistAgentInput {
        client: input.client.clone(),
        resources: Arc::clone(&input.resources),
        github_scope_org: String::new(),
        language: input.language.clone(),
        on_progress_event: Arc::clone(&input.on_progress_event),
        owner_id: LOGS_PROGRESS_OWNER_ID.to_string(),
    });
    let metrics_agent = build_metrics_agent(BuildSpecialistAgentInput {
        client: input.client.clone(),
        resources: Arc::clone(&input.resources),
        github_scope_org: String::new(),
        language: input.language.clone(),
        on_progress_event: Arc::clone(&input.on_progress_event),
        owner_id: METRICS_PROGRESS_OWNER_ID.to_string(),
    });
    let events_agent = build_events_agent(BuildSpecialistAgentInput {
        client: input.client.clone(),
        resources: Arc::clone(&input.resources),
        github_scope_org: String::new(),
        language: input.language.clone(),
        on_progress_event: Arc::clone(&input.on_progress_event),
        owner_id: EVENTS_PROGRESS_OWNER_ID.to_string(),
    });
    let github_agent = build_github_agent(BuildSpecialistAgentInput {
        client: input.client.clone(),
        resources: Arc::clone(&input.resources),
        github_scope_org: input.github_scope_org.clone(),
        language: input.language.clone(),
        on_progress_event: Arc::clone(&input.on_progress_event),
        owner_id: GITHUB_PROGRESS_OWNER_ID.to_string(),
    });

    input
        .client
        .agent(INVESTIGATION_MODEL)
        .name("Coordinator")
        .preamble(&build_coordinator_instructions(
            BuildCoordinatorInstructionsInput {
                datadog_site: input.datadog_site,
                github_scope_org: input.github_scope_org,
                runtime: input.runtime,
                language: input.language,
            },
        ))
        .default_max_turns(COORDINATOR_MAX_TURNS)
        .additional_params(model_additional_params())
        .tool(AggregateDatadogLogsByFacetTool::new(Arc::clone(
            &input.resources,
        )))
        .tool(ListDatadogMetricsCatalogTool::new(Arc::clone(
            &input.resources,
        )))
        .tool(ReportProgressTool::new(ReportProgressToolInput {
            on_progress_event: Arc::clone(&input.on_progress_event),
            owner_id: COORDINATOR_PROGRESS_OWNER_ID.to_string(),
        }))
        .tool(ProgressReportingSubAgentTool::new(
            logs_agent,
            LOGS_PROGRESS_OWNER_ID.to_string(),
            Arc::clone(&input.on_progress_event),
        ))
        .tool(ProgressReportingSubAgentTool::new(
            metrics_agent,
            METRICS_PROGRESS_OWNER_ID.to_string(),
            Arc::clone(&input.on_progress_event),
        ))
        .tool(ProgressReportingSubAgentTool::new(
            events_agent,
            EVENTS_PROGRESS_OWNER_ID.to_string(),
            Arc::clone(&input.on_progress_event),
        ))
        .tool(SearchWebTool::new(Arc::clone(&input.resources)))
        .tool(ProgressReportingSubAgentTool::new(
            github_agent,
            GITHUB_PROGRESS_OWNER_ID.to_string(),
            Arc::clone(&input.on_progress_event),
        ))
        .build()
}

#[must_use]
pub fn build_coordinator_prompt(alert_context: &AlertContext) -> String {
    let investigation_prompt = "Investigate the following user input and respond with the most appropriate investigation or direct answer.
The input may be an alert, request, question, link, or partial context.";
    let trigger_message_section = format!(
        "\n\nTrigger Message: {}",
        alert_context.trigger_message_text
    );
    let thread_context_section = if alert_context.thread_transcript.is_empty() {
        String::new()
    } else {
        format!("\n\nThread Context:\n{}", alert_context.thread_transcript)
    };

    format!("{investigation_prompt}{trigger_message_section}{thread_context_section}")
}
struct BuildSpecialistAgentInput {
    client: openai::Client,
    resources: Arc<InvestigationResources>,
    github_scope_org: String,
    language: String,
    on_progress_event: Arc<dyn InvestigationProgressEventPort>,
    owner_id: String,
}

fn build_logs_agent(input: BuildSpecialistAgentInput) -> OpenAiSubAgent {
    input
        .client
        .agent(INVESTIGATION_MODEL)
        .name("investigate_logs")
        .description("Delegates Datadog log investigation tasks.")
        .preamble(&build_logs_instructions(&input.language))
        .default_max_turns(SUBAGENT_MAX_TURNS)
        .additional_params(model_additional_params())
        .hook(ProgressEventHook::new(
            input.owner_id.clone(),
            Arc::clone(&input.on_progress_event),
        ))
        .tool(ReportProgressTool::new(ReportProgressToolInput {
            on_progress_event: Arc::clone(&input.on_progress_event),
            owner_id: input.owner_id,
        }))
        .tool(SearchDatadogLogsTool::new(Arc::clone(&input.resources)))
        .tool(SearchWebTool::new(input.resources))
        .build()
}

fn build_metrics_agent(input: BuildSpecialistAgentInput) -> OpenAiSubAgent {
    input
        .client
        .agent(INVESTIGATION_MODEL)
        .name("investigate_metrics")
        .description("Delegates Datadog metrics investigation tasks.")
        .preamble(&build_metrics_instructions(&input.language))
        .default_max_turns(SUBAGENT_MAX_TURNS)
        .additional_params(model_additional_params())
        .hook(ProgressEventHook::new(
            input.owner_id.clone(),
            Arc::clone(&input.on_progress_event),
        ))
        .tool(ReportProgressTool::new(ReportProgressToolInput {
            on_progress_event: Arc::clone(&input.on_progress_event),
            owner_id: input.owner_id,
        }))
        .tool(QueryDatadogMetricsTool::new(Arc::clone(&input.resources)))
        .tool(SearchWebTool::new(input.resources))
        .build()
}

fn build_events_agent(input: BuildSpecialistAgentInput) -> OpenAiSubAgent {
    input
        .client
        .agent(INVESTIGATION_MODEL)
        .name("investigate_events")
        .description("Delegates Datadog event investigation tasks.")
        .preamble(&build_events_instructions(&input.language))
        .default_max_turns(SUBAGENT_MAX_TURNS)
        .additional_params(model_additional_params())
        .hook(ProgressEventHook::new(
            input.owner_id.clone(),
            Arc::clone(&input.on_progress_event),
        ))
        .tool(ReportProgressTool::new(ReportProgressToolInput {
            on_progress_event: Arc::clone(&input.on_progress_event),
            owner_id: input.owner_id,
        }))
        .tool(SearchDatadogEventsTool::new(Arc::clone(&input.resources)))
        .tool(SearchWebTool::new(input.resources))
        .build()
}

fn build_github_agent(input: BuildSpecialistAgentInput) -> OpenAiSubAgent {
    input
        .client
        .agent(INVESTIGATION_MODEL)
        .name("investigate_github")
        .description("Delegates GitHub repository, code, and pull request investigation tasks.")
        .preamble(&build_github_instructions(BuildGithubInstructionsInput {
            language: input.language,
            github_scope_org: input.github_scope_org.clone(),
        }))
        .default_max_turns(SUBAGENT_MAX_TURNS)
        .additional_params(model_additional_params())
        .hook(ProgressEventHook::new(
            input.owner_id.clone(),
            Arc::clone(&input.on_progress_event),
        ))
        .tool(ReportProgressTool::new(ReportProgressToolInput {
            on_progress_event: Arc::clone(&input.on_progress_event),
            owner_id: input.owner_id,
        }))
        .tool(SearchGithubCodeTool::new(
            Arc::clone(&input.resources.github_code_search_port),
            input.github_scope_org.clone(),
        ))
        .tool(SearchGithubReposTool::new(
            Arc::clone(&input.resources.github_code_search_port),
            input.github_scope_org.clone(),
        ))
        .tool(SearchGithubIssuesAndPullRequestsTool::new(
            Arc::clone(&input.resources.github_code_search_port),
            input.github_scope_org.clone(),
        ))
        .tool(GetRepositoryContentTool::new(
            Arc::clone(&input.resources.github_repository_content_port),
            input.github_scope_org.clone(),
        ))
        .tool(GetPullRequestTool::new(
            Arc::clone(&input.resources.github_pull_request_port),
            input.github_scope_org.clone(),
        ))
        .tool(GetPullRequestDiffTool::new(
            Arc::clone(&input.resources.github_pull_request_port),
            input.github_scope_org,
        ))
        .tool(SearchWebTool::new(input.resources))
        .build()
}

#[must_use]
fn model_additional_params() -> serde_json::Value {
    json!({
        "reasoning": {
            "effort": "low",
            "summary": "auto",
        },
        "text": {
            "format": {
                "type": "text",
            },
        },
        "parallel_tool_calls": true,
    })
}

struct BuildCoordinatorInstructionsInput {
    datadog_site: String,
    github_scope_org: String,
    runtime: InvestigationRuntime,
    language: String,
}

fn build_coordinator_instructions(input: BuildCoordinatorInstructionsInput) -> String {
    let datadog_site = if input.datadog_site.is_empty() {
        "datadoghq.com".to_string()
    } else {
        input.datadog_site
    };

    format!(
        "You are an SRE/Security/Platform engineer operating from Slack mentions.

Output language: {language}
- Use {language} for all responses and reasoning.

Current run context:
- Now: {now}
- StartedAt: {started_at_iso}
- Slack Channel: {channel}
- Slack Thread: {thread_ts}
- Retry Count: {retry_count}
- GitHub Organization Scope: {github_scope_org}

You orchestrate investigation end-to-end.
- First classify if the request is incident investigation or direct retrieval.
- Before entering a new investigation step, call report_progress.
- report_progress payload must be short: use title and summary fields.
- Do not post consecutive report_progress calls with identical content.
- Your response is posted to Slack as-is. Do not rely on any downstream rewriting.
- Write the final response as a concise, scannable Slack message using Slack markdown.
- For investigation mode, establish system map from GitHub before deep Datadog querying.
- Use aggregate_datadog_logs_by_facet and list_datadog_metrics_catalog early to understand service scope.
- Delegate detailed work to investigate_logs / investigate_metrics / investigate_events / investigate_github as needed.
- Run independent tool calls in parallel where possible.

Web search:
- Use search_web to check whether external dependencies (cloud providers, CDNs, DNS, third-party APIs, SaaS platforms) are experiencing outages or degraded performance that could explain the symptoms observed internally.
- When internal metrics or logs suggest connectivity issues, elevated error rates toward external endpoints, or timeouts on third-party calls, proactively search for recent public incident reports or status page updates for those services.

Final answer requirements:
- In investigation mode, provide several plausible findings with evidence and confidence.
- Include Datadog deep links with this site: {datadog_site}
- Include short sections: What I checked / What I did not find (if relevant).
- In direct task mode, provide concise direct answer with minimal extra content.",
        language = input.language,
        now = Utc::now().to_rfc3339(),
        started_at_iso = input.runtime.started_at_iso,
        channel = input.runtime.channel,
        thread_ts = input.runtime.thread_ts,
        retry_count = input.runtime.retry_count,
        github_scope_org = input.github_scope_org,
        datadog_site = datadog_site,
    )
}

fn build_logs_instructions(language: &str) -> String {
    format!(
        "You are a log analysis specialist.
Before entering a new investigation step, call report_progress.
report_progress payload must be short: use title and summary fields.
Do not post consecutive report_progress calls with identical content.
Use search_datadog_logs and summarize errors, anomalies, and patterns.
Run independent tool calls in parallel when possible.
If you receive client_error payloads, adjust query and retry when useful.
Use {language} for all responses."
    )
}

fn build_metrics_instructions(language: &str) -> String {
    format!(
        "You are a metrics analysis specialist.
Before entering a new investigation step, call report_progress.
report_progress payload must be short: use title and summary fields.
Do not post consecutive report_progress calls with identical content.
Use query_datadog_metrics and summarize trends, spikes, and anomalies.
Run independent tool calls in parallel when possible.
If you receive client_error payloads, adjust query and retry when useful.
Use {language} for all responses."
    )
}

fn build_events_instructions(language: &str) -> String {
    format!(
        "You are an events analysis specialist.
Before entering a new investigation step, call report_progress.
report_progress payload must be short: use title and summary fields.
Do not post consecutive report_progress calls with identical content.
Use search_datadog_events and correlate deployments/config changes with incidents.
Run independent tool calls in parallel when possible.
If you receive client_error payloads, adjust query and retry when useful.
Use {language} for all responses."
    )
}

struct BuildGithubInstructionsInput {
    language: String,
    github_scope_org: String,
}

fn build_github_instructions(input: BuildGithubInstructionsInput) -> String {
    format!(
        "You are a GitHub analysis specialist.
Before entering a new investigation step, call report_progress.
report_progress payload must be short: use title and summary fields.
Do not post consecutive report_progress calls with identical content.
Use the available tools to search code, repositories, issues, pull requests, and repository files.
Mandatory query rule:
- Every search_github_code/search_github_repos/search_github_issues_and_pull_requests call must include org:{github_scope_org}
- Never omit the org qualifier.
Run independent searches in parallel when possible.
Use {language} for all responses.",
        github_scope_org = input.github_scope_org,
        language = input.language,
    )
}

#[cfg(test)]
mod tests {
    use reili_shared::ports::outbound::InvestigationRuntime;
    use reili_shared::types::AlertContext;
    use serde_json::json;

    use super::{
        BuildCoordinatorInstructionsInput, BuildGithubInstructionsInput,
        build_coordinator_instructions, build_coordinator_prompt, build_events_instructions,
        build_github_instructions, build_logs_instructions, build_metrics_instructions,
        model_additional_params,
    };

    fn sample_alert_context() -> AlertContext {
        AlertContext {
            raw_text: "Please investigate this alert".to_string(),
            trigger_message_text: "Please investigate this alert".to_string(),
            thread_transcript: "thread context".to_string(),
        }
    }

    #[test]
    fn builds_coordinator_prompt_with_thread_context() {
        let prompt = build_coordinator_prompt(&sample_alert_context());
        assert!(prompt.contains("Trigger Message: Please investigate this alert"));
        assert!(prompt.contains("Thread Context:\nthread context"));
    }

    #[test]
    fn model_additional_params_enable_parallel_tool_calls() {
        let params = model_additional_params();
        assert_eq!(
            params.get("parallel_tool_calls"),
            Some(&serde_json::Value::Bool(true))
        );
        assert_eq!(
            params.get("text"),
            Some(&json!({ "format": { "type": "text" } }))
        );
    }

    #[test]
    fn coordinator_instructions_include_report_progress_rules() {
        let instructions = build_coordinator_instructions(BuildCoordinatorInstructionsInput {
            datadog_site: "datadoghq.com".to_string(),
            github_scope_org: "acme".to_string(),
            runtime: InvestigationRuntime {
                started_at_iso: "2026-01-01T00:00:00Z".to_string(),
                channel: "C123".to_string(),
                thread_ts: "123.456".to_string(),
                retry_count: 0,
            },
            language: "Japanese".to_string(),
        });

        assert!(instructions.contains("call report_progress"));
        assert!(instructions.contains("title and summary fields"));
        assert!(instructions.contains("Do not post consecutive report_progress"));
    }

    #[test]
    fn specialist_instructions_include_report_progress_rules() {
        let logs_instructions = build_logs_instructions("Japanese");
        let metrics_instructions = build_metrics_instructions("Japanese");
        let events_instructions = build_events_instructions("Japanese");
        let github_instructions = build_github_instructions(BuildGithubInstructionsInput {
            language: "Japanese".to_string(),
            github_scope_org: "acme".to_string(),
        });

        for instructions in [
            logs_instructions,
            metrics_instructions,
            events_instructions,
            github_instructions,
        ] {
            assert!(instructions.contains("call report_progress"));
            assert!(instructions.contains("title and summary fields"));
            assert!(instructions.contains("Do not post consecutive report_progress"));
        }
    }

    #[test]
    fn coordinator_instructions_include_web_search_rules() {
        let instructions = build_coordinator_instructions(BuildCoordinatorInstructionsInput {
            datadog_site: "datadoghq.com".to_string(),
            github_scope_org: "acme".to_string(),
            runtime: InvestigationRuntime {
                started_at_iso: "2026-01-01T00:00:00Z".to_string(),
                channel: "C123".to_string(),
                thread_ts: "123.456".to_string(),
                retry_count: 0,
            },
            language: "Japanese".to_string(),
        });

        assert!(instructions.contains("search_web"));
        assert!(instructions.contains("external dependencies"));
        assert!(instructions.contains("public incident reports or status page"));
    }
}
